upgrade/upgrade-scripts/upgrade

```diff
@@ -395,16 +397,21 @@ function finalize() {
 		;;
 	esac
 
-	ROOTFS_DATASET=$(get_mounted_rootfs_container_dataset)
-	[[ -n "$ROOTFS_DATASET" ]] ||
-		die "unable to determine mounted rootfs container dataset"
-
-	SNAPSHOT_NAME=$(get_dataset_rollback_snapshot_name "$ROOTFS_DATASET")
-	[[ -n "$SNAPSHOT_NAME" ]] ||
-		die "unable to determine rollback snapshot name"
-
-	zfs destroy -r "$ROOTFS_DATASET@$SNAPSHOT_NAME" ||
-		die "failed to destroy rollback snapshots"
+	SNAPSHOT_LIST=$(zfs list -t snapshot -r rpool/ROOT -d 2 -Ho name) ||
+		die "unable to determine available rollback snapshots"
+
+	for snapshot in $SNAPSHOT_LIST; do
+		grep -qE "^execute-upgrade.[[:alnum:]]{7}$" <(echo "$snapshot" | awk -F@ '{print $2}') ||
+			die "unexpected snapshot name: '$snapshot'"
+		local SNAPSHOT_VERSION
+		SNAPSHOT_VERSION=$(zfs get -Hpo value "$PROP_CURRENT_VERSION" "$snapshot") ||
+			die "failed to get snapshot version"
+		if compare_versions \
+			"$SNAPSHOT_VERSION" "lt" "$UPGRADE_BASE_VERSION"; then
```

Comment:
It's worth noting that.. since we're comparing versions here, if we end up "upgrading" to an older version, we won't end up removing the snapshots of that previous version..

e.g. I can simulate this by cloning a trunk VM, and then "upgrading" to a upgrade image that was generated earlier than that trunk VM.. after doing something like that, I wind up with datasets looking like this:
```
rpool/ROOT/delphix.SVehJOD                               com.delphix:current-version  6.1.0.0-snapshot.20200310170304292+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-4447  local
rpool/ROOT/delphix.SVehJOD@execute-upgrade.NXkUlTF       com.delphix:current-version  6.1.0.0-snapshot.20200311085025156+jenkins-ops-devops-gate-master-appliance-build-master-post-push-2728         local
```

i.e. `6.1.0.0-snapshot.20200311085025156+jenkins-ops-devops-gate-master-appliance-build-master-post-push-2728` was the original version of the trunk VM before upgrade.. then I upgraded using an upgrade image containing version `6.1.0.0-snapshot.20200310170304292+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-4447`

After performing the upgrade, I was now running version `6.1.0.0-snapshot.20200310170304292+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-4447`, but since that version number is actually less than what I started with.. when I later call `finalize`, it won't remove the snapshot.
 
I'm not sure it's worth changing the code you have here to account for such a use case, since I don't think a customer system will ever run into this issue (i.e. I don't think we'll ever upgrade to a version that's less than what's currently running, on a customer's system), but I do think it's something to be aware of..

Reply:
@prakashsurya I believe the API route of upgrade does not allow for upload and unpack of older version image. Was not aware that the scripts provide a way to do this which is good to know for some internal testing situations.

Reply:
:+1: OK, good to know.

> Was not aware that the scripts provide a way to do this

Yea, the underlying logic (i.e. the scripts) try to allow "upgrading" to any upgrade image, no matter the "from" and "to" versions. This was intentional, so the logic to implement policy (e.g. one can only upgrade to a "newer" version) could be implemented at the higher layers; which sounds like exactly what we're doing in the API, so that's good to know.